digdag-core/src/main/java/io/digdag/core/agent/ExtractArchiveWorkspaceManager.java

```diff
@@ -22,19 +31,53 @@
     public ExtractArchiveWorkspaceManager(TempFileManager tempFiles)
     {
         this.tempFiles = tempFiles;
+
     }
 
     @Override
     public <T> T withExtractedArchive(TaskRequest request, ArchiveProvider archiveProvider, WithWorkspaceAction<T> func)
             throws IOException
     {
-        try (TempDir workspacePath = createNewWorkspace(request)) {
-            Optional<StorageObject> in = archiveProvider.open();
-            if (in.isPresent()) {
-                ProjectArchives.extractTarArchive(workspacePath.get(), in.get().getContentInputStream());
+        TempDir workspacePath = null;
+        try {
+            try {
+                // A temporary workspace should be created for each try
+                // And always close the temporary workspace
+                workspacePath = RetryExecutor.retryExecutor()
+                        .retryIf(exception -> true)
+                        .withInitialRetryWait(EXTRACT_MIN_RETRY_WAIT_MS)
+                        .withMaxRetryWait(EXTRACT_MAX_RETRY_WAIT_MS)
+                        .onRetry((exception, retryCount, retryLimit, retryWait) ->
+                                logger.warn("Failed to extract archive: retry {} of {}", retryCount, retryLimit, exception))
+                        .withRetryLimit(EXTRACT_RETRIES)
+                        .run(() -> {
+                            TempDir newWorkSpacePath = null;
+                            try {
+                                newWorkSpacePath = createNewWorkspace(request);
+                                Optional<StorageObject> in = archiveProvider.open();
+                                if (in.isPresent()) {
+                                    ProjectArchives.extractTarArchive(newWorkSpacePath.get(), in.get().getContentInputStream());
+                                }
+                                return newWorkSpacePath;
+                            }
+                            catch (Throwable e) {
```

Comment:
you can use `finally` here and remove `throw e` at the end.

Reply:
With `finally` here, returned `TempDir` from here to outside RetryExecutor#run will be always closed and below `func.run` can't use, I think.

Reply:
In the other words, this `close` is only for when `ProjectArchives.extractTarArchive` fails

Reply:
doesn't this work?:
```
}
finally {
    if (newWorkSpacePath != null) {
        newWorkSpacePath.close();
    }
}
```

Reply:
oh no. it doesn't work. now I got it....!